core/templates/dev/head/forms/formBuilder.js

```diff
@@ -747,9 +726,383 @@ oppia.config(['$provide', function($provide) {
+        // Icon for each component.
+        var icons = [];
+        // CKEditor plugin/widget name for each component.
+        var pluginNames = [];
+        _RICH_TEXT_COMPONENTS.forEach(function(componentDefn) {
+          if (!((scope.uiConfig() &&
+                 scope.uiConfig().hide_complex_extensions &&
+                 componentDefn.isComplex) ||
+                 (!canUseFs && componentDefn.requiresFs))) {
+            componentNames.push(componentDefn.name);
+            icons.push(componentDefn.iconDataUrl);
+            pluginNames.push(componentDefn.ckWidgetName);
+          }
+        });
+
+        /**
+         * Create rules to whitelist all the rich text components and
+         * their wrappers and overlays.
+         * See format of filtering rules here:
+         * http://docs.ckeditor.com/#!/guide/dev_allowed_content_rules
+         */
+        // Whitelist the component tags with any attributes and classes.
+        var componentRule = componentNames.map(function(name) {
+          return 'oppia-noninteractive-' + name;
+        }).join(' ') + '(*)[*];';
+        // Whitelist the inline component wrapper, which is a
+        // span with a "type" attribute.
+        var inlineWrapperRule = ' span[type];';
+        // Whitelist the block component wrapper, which is a div
+        // with a "type" attribute and a CSS class.
+        var blockWrapperRule = ' div(oppia-rte-component-container)[type];';
+        // Whitelist the transparent block component overlay, which is
+        // a div with a CSS class.
+        var blockOverlayRule = ' div(oppia-rte-component-overlay);';
+        // Put all the rules together.
+        var extraAllowedContentRules = componentRule +
+                                       inlineWrapperRule +
+                                       blockWrapperRule +
+                                       blockOverlayRule;
+
+        // The button corresponding to a widget has the same name,
+        // with a capitalized first letter.
+        var buttonNames = [];
+        pluginNames.forEach(function(pluginName) {
+          buttonNames.push(pluginName.charAt(0).toUpperCase() +
+                           pluginName.slice(1, pluginName.length));
+          // '-' adds the vertical divider between toolbar buttons.
+          buttonNames.push('-');
+        });
+        // Remove the last '-' from the list.
+        buttonNames.pop();
+
+        // Initialize ckeditor.
+        var ck = CKEDITOR.inline(el[0].children[0].children[1], {
+          extraPlugins: 'widget,lineutils,sharedspace,' + pluginNames.join(','),
+          startupFocus: true,
+          extraAllowedContent: extraAllowedContentRules,
+          sharedSpaces: {
+            // Select the directive template's empty div.
+            top: el[0].children[0].children[0]
+          },
+          skin: 'bootstrapck,/third_party/static/ckeditor-bootstrapck/',
+          toolbar: [
+            {
+              name: 'history',
```

Comment:
Just to check, are these names arbitrary? If so, might be worth mentioning this.